charms/istio-pilot/src/charm.py

```diff
@@ -441,16 +444,38 @@ def _check_leader(self):
             raise ErrorWithStatus("Waiting for leadership", WaitingStatus)
 
     @property
-    def _cert_subject(self):
-        """Return the domain to be used in the CSR."""
+    def _ingress_gateway_host(self) -> Optional[str]:
+        """Return a domain name if it's configured, the ingress gateway svc address otherwise."""
+        # If the configuration option is set, we use it
+        if domain_name := self.model.config["domain-name"]:
+            return domain_name
+
+        # We try getting the ingress gateway IP address in the absence of domain name
         try:
             svc = self._get_gateway_service()
         except ApiError:
-            self.log.info("Could not retrieve the gateway service for configuring the CSR.")
+            self.log.info("Could not retrieve the gateway service for using in the CSR.")
             return None
-        gateway_address = _get_gateway_address_from_svc(svc)
-        if gateway_address:
-            return gateway_address
+        return _get_gateway_address_from_svc(svc)
+
+    @property
+    def _cert_subject(self) -> Optional[str]:
+        """Return the certificate subject to be used in the CSR."""
+        target = self._ingress_gateway_host
+
+        # Check if the target is a hostname and return if it is
+        if _is_hostname(target):
+            assert isinstance(target, str), target
+            return target
+
+        # If target is an IP address, try to get the hostname and return it
+        with contextlib.suppress(OSError, TypeError):
+            name, _, _ = socket.gethostbyaddr(target)
+            if _is_hostname(name) and not name.endswith(".svc.cluster.local"):
```

Comment:
`_is_hostname(name)` can never be `True` here, because 5 lines above this we have `if is_hostname(target): ... return`